### ISSUE TITLE
FR: add auto-update mechanism

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#755]
We need a Windows auto-updater. Windows updates the slowest of all our platforms:

<img width="1071" alt="Screen Shot 2020-09-11 at 12 19 00 PM" src="https://user-images.githubusercontent.com/2621/92964761-80cd5c00-f429-11ea-83c6-73e17c2a26e6.png">

Misc notes copied from internal bug:

* It might be opt-in or might be opt-out, and should probably have a registry-settable override for org admins to prevent their managed machines from auto-updating outside the org's schedules. But we should support an auto-update mechanism in any case, even if not everybody wants to use it.
* Checkbox in systray menu, like: `[X] Auto-update software`
* Windows service uses BITS via the bitsadmin CLI tool to download the new installer. It'd be easy in Go too, of course, but too easy to do poorly. By using BITS, we avoid being rude to their network (using too much bandwidth or money) and can instead let Windows decide how quickly and when to download, based on which wi-fi networks they've classified in which ways (marking an LTE hotspot as expensive, for example).
* Need to verify the download once it's downloaded from BITS: https://twitter.com/bradfitz/status/1289300026200481792
* When the download is complete, we simply run the NSIS installer in headless mode, the same way that Chocolatey does it, with /S: https://github.com/zombiezen/tailscale-chocolatey/blob/master/tools/chocolateyinstall.ps1#L20
* The install stops & re-starts everything.
* We can have a intent state file that makes sure we don't try repeatedly in a loop, lest an update go poorly
* we could have the service send a Notify to the GUI to schedule itself to start again when it notices the mtime/inode/size of the GUI is different. So then the GUI can fork off a new process (a copy of the exe to a temp file?) that's polling the file every second for up to 30 seconds and doing the RunFromProcess.exe trick when it notices a new binary is in place? But then it's the GUI restarting itself (upon command) from its context, rather than the admin context.

**Comment by jucor** [comment]
Naive related question: it seems the [official windows installer](https://tailscale.com/download) is stuck at 1.0.4, whereas 1.0.5 is available on Linux and MacOS. Is that intended?
Would *love* an auto-update :)

**Comment by bradfitz** [comment]
There were only two changes in 1.0.5: one was Linux only, and one was a client bug for all platforms that we ended up also working around server-side. So there was no reason to do 1.0.5 on Windows.

**Comment by bradfitz** [comment]
@alexbrainman, this something you'd be interested in working on?

**Comment by alexbrainman** [comment]
> @alexbrainman, this something you'd be interested in working on?

Sure. I would be interested.

> * Checkbox in systray menu, like: `[X] Auto-update software`

I did not look at the code, but could do that.

> * Windows service uses BITS via the bitsadmin CLI tool to download the new installer.

I never heard about BITS. But I can play with bitsadmin command, and make it do what we need from GUI.

> * Need to verify the download once it's downloaded from BITS: https://twitter.com/bradfitz/status/1289300026200481792

I do not understand what you are proposing here. Please explain.

> * When the download is complete, we simply run the NSIS installer in headless mode,

Can do.

> * The install stops & re-starts everything.

Can do.

> * We can have a intent state file that makes sure we don't try repeatedly in a loop, lest an update go poorly

What exactly did you have in mind?

> * we could have the service send a Notify to the GUI to schedule itself to start again when it notices the mtime/inode/size of the GUI is different. So then the GUI can fork off a new process (a copy of the exe to a temp file?) that's polling the file every second for up to 30 seconds and doing the RunFromProcess.exe trick when it notices a new binary is in place? But then it's the GUI restarting itself (upon command) from its context, rather than the admin context.

I did not get it. Explain it slowly this time. ;-)

Thank you.

Alex

**Comment by bradfitz** [comment]
> > Need to verify the download once it's downloaded from BITS: https://twitter.com/bradfitz/status/1289300026200481792
>
> I do not understand what you are proposing here. Please explain.

You can ignore this for now. We can plug in the verification later between the download & the installer exec.

> I did not get it. Explain it slowly this time. ;-)

The problem is that the service needs to run the installer, but the installer is running as the system service and can't interact with the desktop.

But we want the GUI to restart too (if it's running).

So what we'd like to do, prior to re-installing, is have the system Service process send the GUI a message (a "Notify" JSON message over its localhost TCP connection to `ipnserver`) to tell the GUI to stop & restart when the re-install is complete. It would have to start a new process (either our own exe or a PowerShell script) that polls the `tailscale-ipn.exe` and waits for its size/modtime/etc to change, and when it's changed, runs it again. Then we tell the service we did that, stop ourselves, and then the service can do the upgrade and replace the files. The powershell/etc program then notices that `tailscale-ipn.exe` GUI binary is updated and re-runs it (as the user's credentials that can access the desktop).

**Comment by alexbrainman** [comment]
> The problem is that the service needs to run the installer, but the installer is running as the system service and can't interact with the desktop.
> 
> But we want the GUI to restart too (if it's running).
> 
> So what we'd like to do, prior to re-installing, is have the system Service process send the GUI a message (a "Notify" JSON message over its localhost TCP connection to `ipnserver`) to tell the GUI to stop & restart when the re-install is complete.

I understand that. Both service and GUI use the same executable. So they need to be restarted together.

But how do you propose to do the restarting?

Perhaps with GUI we can have an extra parent process that runs GUI. The parent can restart GUI. But the parent should use different executable file from what we are updating, otherwise we won't be able to replace the executable. Or maybe we can replace executable while it is running - we do something similar when building and installing go.exe. But we still won't be able to restart parent process. But maybe we can delay parent process restarting until GUI is restarted by the user (reboot or any other reason).

But I don't see who will restart service. Do you propose to have similar parent process that runs as a service, but all work is done by a child process? But then child cannot be a Windows Service, it should be just a normal process. And I do not know, if Windows Service can run child processes without any restrictions.

Perhaps you meant something different. I can google to see how other projects do it.

Let me know.

Thank you.

Alex

**Comment by bradfitz** [comment]
The NSIS installer already supports stopping both the service & GUI and re-starting it all. We'd mostly do the same thing, with slight tweaks because we the installer would be running as the service without desktop access.

Before we went to run the installer (via its `/S` silent automatic mode), we'd do the thing I said above:

> So what we'd like to do, prior to re-installing, is have the system Service process send the GUI a message (a "Notify" JSON message over its localhost TCP connection to ipnserver) to tell the GUI to stop & restart when the re-install is complete. It would have to start a new process (either our own exe or a PowerShell script) that polls the tailscale-ipn.exe and waits for its size/modtime/etc to change, and when it's changed, runs it again. Then we tell the service we did that, stop ourselves, and then the service can do the upgrade and replace the files. The powershell/etc program then notices that tailscale-ipn.exe GUI binary is updated and re-runs it (as the user's credentials that can access the desktop).

It's the GUI that would start a new start-me-back-up process. I'd proposed cmd.exe/powershell but it could also just copy its exe to a temp file and run "itself" (under a new exe name) with some `--poll-for-new-version-and-restart` flag that when it noticed the original `.exe` location change contents, re-start the GUI.

**Comment by alexbrainman** [comment]
> The NSIS installer already supports stopping both the service & GUI and re-starting it all. We'd mostly do the same thing, with slight tweaks because we the installer would be running as the service without desktop access.

Thanks for explaining Brad. I understand general idea.

Alex

**Comment by DentonGentry** [comment]
Looking into automatic updates for Windows applications, some references:
- Omaha, as used with Chrome (https://omaha-consulting.com/best-update-framework-for-windows-10-7 provides a hosted version)
- [Squirrel](https://github.com/Squirrel/Squirrel.Windows)
- [WinSparkle](https://winsparkle.org/) derived from the MacOS Sparkle
- [wireguard-windows](https://git.zx2c4.com/wireguard-windows/tree/updater) update service

**Comment by archon810** [comment]
Any updates on this? There are lots of requests for this feature, and updating a lot of machines manually is a management pain.

**Comment by DentonGentry** [comment]
We're proceeding with Sparkle for the MacOS client, which will mean we'll set up the AppCast RSS feed to trigger Sparkle to update.

That makes [WinSparkle](https://winsparkle.org/) pretty compelling, as I'd expect it to be straightforward to generate another set of AppCast entries for the Windows client.

Of the other options:
- Though Omaha works well for the Windows Chrome, it is reportedly very painful to operate in production.
- [Squirrel](https://github.com/Squirrel/Squirrel.Windows) is now unmaintained. Were we using it we'd probably keep using it, but I don't think it is wise to start using it now.
- The [Wireguard Windows](https://git.zx2c4.com/wireguard-windows/tree/updater) has potential, but if we've already committed to Sparkle for MacOS then WinSparkle seems to have more potential for us.

**Comment by DentonGentry** [comment]
From engineering meeting yesterday:
- After looking into the implementation needs in the app, WinSparkle has some downsides like no ARM support. Its biggest advantage was that we could leverage the AppCast RSS updates, but it turns out that WinSparkle uses a different (older) version of the RSS feed format which macOS no longer uses.
- The update service which the Wireguard project developed looks better from an implementation perspective.

**Comment by clrxbl** [comment]
Is there any update on this? With recent breaking updates like https://tailscale.com/blog/windows-updates/, manually asking users to update their Tailscale version is a hassle.

**Comment by AdrianFletcher** [comment]
Another plus one on this request. It may also be useful to utilise something like update 'rings' or 'channels', such as:

- Unstable
- Release
- Delayed (e.g. install the latest version, but wait until it has been out for at least X number of days)

**Comment by DentonGentry** [comment]
https://git.zx2c4.com/wireguard-windows/tree/updater is the most likely candidate for this functionality.

**Comment by ajnadox** [comment]
Thanks for pointing me to this thread :) as previous posters i also see this as an high prio case for usability and as i wrote when check for new version is made today why not just launch a new process that updates it to the new version at the same time, then when user reboots computer or when tailscale is idle, the process with old tailscale is killed and new launched, update dont need to be instant just so we know atleast a reboot of computer will update it.

**Comment by wavded** [comment]
We ran into this same issue with my team. Appreciate that this issue exists and the problem is known and you all are looking into it!

**Comment by freezscholte** [comment]
@DentonGentry any updates on this?

**Comment by ukanuk** [comment]
The actual automatic update is a lot of work to implement, I get it.

In the meantime, could Tailscale on Windows check for updates and send a notification for users to manually update?

My main issue is I never even think to check for updates, and I'll randomly remember to update and be months and months out of date.

**Comment by bradfitz** [comment]
@dblohm7 wrote on #6428:

> How should we solve this?
> Prereq: No more NSIS installer. 100% MSI.
> Needs changes both on client and server-side (probably pkgs)

**Comment by loosus** [comment]
I am going to add one suggestion to this. That's easy for me to use as an end user who doesn't do the work. :)

I know the initial deployment of an auto updater might not have these features, but it might be worth thinking about Group Policies and MDM policies to control the updater as it is developed.  Policies like deferring updates by X number of days from an update's release date, getting updates from a third-party HTTPS server, indicating the day of the week and approximate time of day the installation can occur, etc.  Developing an updater with this future in mind might make adding these features later easier.

I would think Omaha would have some of this already built-in, but I've heard people say Omaha is a bear to get up and running.

**Comment by DentonGentry** [comment]
I don't think WinSparkle is likely at this point, its development is essentially abandoned.

Omaha does handle just about everything one could want. However as you say, operating it is _hard_. That is the kind of system one uses when one is Google.

We additionally want to get the app listed in the Microsoft Windows Store, which has its own update mechanism that will need to be accommodated.

**Comment by DentonGentry** [comment]
In Tailscale 1.36.0 though the NSIS installer is still built, it is not linked from pkgs.tailscale.com and the installer has moved to MSI. It is expected to be easier to autoupdate the MSI.

**Comment by loosus** [comment]
That's a good move.

How are the CLI updater and the transition to MSI related?

I mean, could I create a scheduled task that runs as SYSTEM that runs the updater?

**Comment by DentonGentry** [comment]
I understand the desire to have this issue addressed. We do bug scrubs every few weeks, and attempt to update the current status of all open bugs wherever the status has changed.

The `tailscale update` command and the switch to MSI are both related to this bug, but do not constitute a completed implementation which you will be able to use right now.

**Comment by loosus** [comment]
Got it.  So just running `tailscale update --yes` as SYSTEM would not perform the desired action?

**Comment by DentonGentry** [comment]
Run manually from the CLI or some sort of automation, `tailscale update --yes` will update to the current stable version. This is considered to be an Alpha feature, but does work on Windows: https://tailscale.com/kb/1080/cli/#update-alpha

Right now nothing within Tailscale will automatically run `tailscale update` periodically to keep the software updated, and it doesn't work on platforms other than Windows yet.

**Comment by loosus** [comment]
At least for Windows, that sounds dangerously close to complete. :)

Just to give some context for my own organization: the only place we use Linux extensively is on our servers, and updating there isn't terribly difficult even without an updater because servers are so captive.  In a worst-case scenario, I could manually login to them and have them all updated in 60 minutes.  Not a huge deal.

Where we have the most trouble are our out-in-the-field users who could be located anywhere on Earth.  Those are all 100% Windows 10 or Windows 11 devices.  These are the ones that (a) the IT department has little access to outside of a management tool like Intune and (b) users don't have admin rights to.

So if I can just create a scheduled task that runs at, say, 1 AM that tells Tailscale to update, that's almost all I need to keep our out-in-the-field users updated.  And I could conceivably create a desktop shortcut that runs the scheduled task so users could manually update Tailscale in case their devices are not turned on at the scheduled update time (e.g., 1 AM).  Sidenote: I wonder if I could create a separate scheduled task that does nothing but checks the currently installed version of Tailscale and compares it against the latest available version, and if found to be outdated would somehow inform the user?

I am not trying to discourage you from further GUI work on the updater, but that CLI updater alone would probably go a long ways for many desperate IT departments.

**Comment by loosus** [comment]
Some feedback on using `tailscale update -yes=true` on Windows: it seems to work but it hangs forever at the end.

```
C:\Windows\system32>tailscale update -yes=true
2023/02/26 19:05:05 Updating Tailscale from 1.36.1 to 1.36.2; --yes given, continuing without prompts.
2023/02/26 19:05:05 Download size: 19003392
2023/02/26 19:05:06 Downloaded 16384/19003392 (0.1%)
2023/02/26 19:05:07 Downloaded 19003392/19003392 (100.0%)
2023/02/26 19:05:07 hash matched
2023/02/26 19:05:07 verifying MSI authenticode...
2023/02/26 19:05:07 authenticode verification succeeded
2023/02/26 19:05:07 making tailscale.exe copy to switch to...
2023/02/26 19:05:07 running tailscale.exe copy for final install...

C:\Windows\system32>2023/02/26 19:05:08 installing C:\ProgramData\Tailscale\MSICache\tailscale-setup-1.36.2-amd64.msi ...
2023/02/26 19:05:15 success.
```

**Comment by dandill-curative** [comment]
Just mentioning that this would be a handy client side feature, which would improve adoption of updated versions. I've found that other applications that allow for enabling client auto-update but an admin specified 'highest version to update to' also helps for release management purposes.

**Comment by filviu** [comment]
Hangs for me too. Windows 11. Upgrade from 1.40 to 1.40.1

**Comment by awly** [comment]
Auto-update Beta is now available in 1.52: https://tailscale.com/blog/auto-update-beta/
I'll close this, we can open more specific issues as needed.
